### PR TITLE
feat: [ANDROAPP-2732] New section navigation for events

### PR DESCRIPTION
**Pull request by Balcan**
[issue](https://jira.dhis2.org/browse/ANDROAPP-2732)

### CODE REVIEW

File: app/src/main/java/org/dhis2/data/forms/dataentry/DataEntryAdapter.java [hunk 1]
```diff
@@ -316,23 +339,33 @@ public boolean hasError() {
...
+        if (getItemViewType(position) == SECTION || rendering == null) {
+            return 2;
+        } else {
+            switch (ProgramStageSectionRenderingType.valueOf(rendering)) {
+                case MATRIX:
+                    return 1;
+                case LISTING:
+                default:
+                case SEQUENTIAL:
+                    return 2;
```
Comment (ferdyrod) [inline app/src/main/java/org/dhis2/data/forms/dataentry/DataEntryAdapter.java]: It looks a little weird to have the default case in the middle of other cases and not at the end.

File: app/src/main/java/org/dhis2/data/forms/dataentry/EnrollmentRepository.java [hunk 2]
```diff
@@ -53,9 +58,36 @@ public EnrollmentRepository(@NonNull Context context,
...
+                                return Single.just(fields);
+                            }
+                        })
+                ).toFlowable();
+
+     /*   return d2.enrollmentModule().enrollments().uid(enrollmentUid).get()
                 .flatMap(enrollment -> d2.programModule().programTrackedEntityAttributes().withRenderType().byProgram().eq(enrollment.program()).get()).toFlowable()
                 .flatMapIterable(programTrackedEntityAttributes -> programTrackedEntityAttributes)
-                .map(this::transform).toList().toFlowable();
+                .map(this::transform).toList().toFlowable();*/
```
Comment (ferdyrod) [inline app/src/main/java/org/dhis2/data/forms/dataentry/EnrollmentRepository.java]: Is this needed?

File: app/src/main/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/EventCaptureModule.java [hunk 3]
```diff
@@ -16,6 +16,8 @@
 import org.hisp.dhis.android.core.D2;
 import org.hisp.dhis.rules.RuleExpressionEvaluator;
 
+import javax.inject.Singleton;
+
```
Comment (ferdyrod) [inline app/src/main/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/EventCaptureModule.java]: Unnecessary import

File: app/src/main/java/org/dhis2/utils/customviews/CoordinatesView.java [hunk 4]
```diff
@@ -306,6 +306,15 @@ public void onFocusChange(View v, boolean hasFocus) {
     @SuppressLint("MissingPermission")
     public void updateLocation(Geometry geometry) {
 
+        setCoordinatesValue(geometry);
+
+        this.currentGeometry = geometry;
+        if (listener2 != null)
+            listener2.onCurrentLocationClick(geometry);
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/utils/customviews/CoordinatesView.java]: I know it was here, but if you can change the name listener2 please

File: app/src/main/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/EventCaptureActivity.java [hunk 5]
```diff
@@ -256,8 +232,8 @@ public void setProgramStage(String programStageUid) {
 
     @Override
     public void showRuleCalculation(Boolean shouldShow) {
-        if(binding.eventTabLayout.getSelectedTabPosition() == 0) {
-            binding.calculationIndicator.getRoot().setVisibility(shouldShow ? View.VISIBLE : View.GONE);
+        if (binding.eventTabLayout.getSelectedTabPosition() == 0) {
+//            binding.calculationIndicator.getRoot().setVisibility(shouldShow ? View.VISIBLE : View.GONE);
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/EventCaptureActivity.java]: ???

File: app/src/main/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/EventCaptureFragment/EventCaptureFormPresenter.kt [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            view.dataEntryFlowable()
+                .onBackpressureBuffer()
+                .distinctUntilChanged()
+                .subscribeOn(schedulerProvider.io())
+                .observeOn(schedulerProvider.ui())
+                .switchMap { action ->
+                    if (action.lastFocusPosition() != null && action.lastFocusPosition() >= 0) {
+                        this.lastFocusItem = action.id()
+                    }
+                    //TODO: REMOVE FROM EMPTY MANDATORY FIELDS IF EXIST
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/EventCaptureFragment/EventCaptureFormPresenter.kt]: what about this TODO?

File: app/src/main/java/org/dhis2/utils/customviews/CoordinatesView.java (hunk 4 above)
Comment (Balcan) [reply app/src/main/java/org/dhis2/utils/customviews/CoordinatesView.java]: Yes, good idea

File: app/src/main/java/org/dhis2/utils/customviews/CustomTextView.java [hunk 7]
```diff
@@ -94,10 +94,10 @@ else if (isBgTransparent && isLongText)
         descriptionLabel = binding.getRoot().findViewById(R.id.descriptionLabel);
         dummy = findViewById(R.id.dummyFocusView);
 
-        editText.setOnTouchListener((v, event) -> {
+       /* editText.setOnTouchListener((v, event) -> {
             activate();
             return true;
-        });
+        });*/
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/utils/customviews/CustomTextView.java]: ????

File: app/src/main/java/org/dhis2/data/forms/dataentry/EnrollmentRepository.java [hunk 8]
```diff
@@ -53,9 +58,31 @@ public EnrollmentRepository(@NonNull Context context,
...
+                                                        .byProgram().eq(program.uid())
+                                                        .byTrackedEntityAttribute().eq(attribute.uid())
+                                                        .one().blockingGet();
+                                        fields.add(transform(programTrackedEntityAttribute));
+                                    }
+                                }
+                                return Single.just(fields);
+                            }
+                        })
+                ).toFlowable();
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/data/forms/dataentry/EnrollmentRepository.java]: could you extract some of this logic to other method/class so we can add unit test?

File: app/src/main/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/EventCaptureActivity.java (hunk 5 above)
Comment (Balcan) [reply app/src/main/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/EventCaptureActivity.java]: :S Don't know why i left it that way... will remove it

File: app/src/main/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/EventCaptureFragment/EventCaptureFormPresenter.kt (hunk 6 above)
Comment (Balcan) [reply app/src/main/java/org/dhis2/usescases/eventsWithoutRegistration/eventCapture/EventCaptureFragment/EventCaptureFormPresenter.kt]: Already done! :) Removing it as well

File: app/src/main/java/org/dhis2/data/forms/dataentry/DataEntryAdapter.java [hunk 9]
```diff
@@ -36,22 +36,28 @@
...
 
 import java.util.ArrayList;
 import java.util.List;
 
 import io.reactivex.processors.FlowableProcessor;
 import io.reactivex.processors.PublishProcessor;
-import timber.log.Timber;
 
-public final class DataEntryAdapter extends Adapter {
+public final class DataEntryAdapter extends ListAdapter<FieldViewModel, ViewHolder> /*implements StickyHeaderListener*/ {
```
Comment (JaimeToca) [inline app/src/main/java/org/dhis2/data/forms/dataentry/DataEntryAdapter.java]: remove comment?

File: app/src/main/java/org/dhis2/utils/customviews/CustomTextView.java (hunk 7 above)
Comment (Balcan) [reply app/src/main/java/org/dhis2/utils/customviews/CustomTextView.java]: Done! ;)

File: app/src/main/java/org/dhis2/data/forms/dataentry/EnrollmentRepository.java (hunk 8 above)
Comment (JaimeToca) [reply app/src/main/java/org/dhis2/data/forms/dataentry/EnrollmentRepository.java]: or add unit test to this method in the class?

Comment (JaimeToca) [reply app/src/main/java/org/dhis2/data/forms/dataentry/EnrollmentRepository.java]: thanks!

Comment (Balcan) [reply app/src/main/java/org/dhis2/data/forms/dataentry/EnrollmentRepository.java]: I'll create two public methods with the VisisbleForTesting tag

Comment (JaimeToca) [reply app/src/main/java/org/dhis2/data/forms/dataentry/EnrollmentRepository.java]: I think you don't need it and you can tests the whole method?

Comment (JaimeToca) [reply app/src/main/java/org/dhis2/data/forms/dataentry/EnrollmentRepository.java]: visible for testing is not usually a good practise, I use it for only exceptions..

File: app/src/main/java/org/dhis2/data/forms/dataentry/fields/section/SectionHolder.kt [hunk 10]
```diff
@@ -0,0 +1,116 @@
...
+        this.viewModel = viewModel
+        formBinding.apply {
+            sectionName.text = viewModel.label()
+            openIndicator.visibility = if (viewModel.isOpen) View.VISIBLE else View.GONE
+            sectionFieldsInfo.text = String.format(
+                "%s/%s",
+                viewModel.completedFields(),
+                viewModel.totalFields()
+            )
+        }
```
[the remark this answers is not in the document]
Comment (Balcan) [reply app/src/main/java/org/dhis2/data/forms/dataentry/fields/section/SectionHolder.kt]: lint does not like that. I'll live it as it is now

**Pull request merged**